mt7620: workaround jal imm26 and redundant PAGE_ALIGN

namiltd opened the pull request:
On MT7620-class platforms (CONFIG_NET_RALINK_MT7620) we observe sporadic wrong-jump-targets, kernel oopses, hanging, corrupted backtraces or even "half-written" instructions when the compiler emits a direct 'jal imm26' call. 
This is triggered in:
  - the small random helpers inside get_random_u32_below(), and
  - the blkcg_maybe_throttle_current() call in resume_user_mode_work().

This patch forces those two call sites to use an indirect call via a volatile function pointer (load into register + jalr) when building for MT7620, avoiding embedding a 26-bit immediate jump target.

Additionally, on MT7620 builds the exec path in fs/exec.c is modified:
  - skip arch_align_stack() + PAGE_ALIGN() in setup_arg_pages() because the micro-randomization (< PAGE_SIZE) implemented by many ports (including MT7620) is negated immediately by PAGE_ALIGN(). Skipping the redundant PAGE_ALIGN() reduces exposure to the problematic code pattern.

These changes are targeted workarounds for MT7620; behavioral logic is unchanged.

Since CONFIG_SOC_MT7620 is also set for mt7628, CONFIG_NET_RALINK_MT7620 is used to detect the correct platform.

Fixes: https://github.com/openwrt/openwrt/issues/16396, https://github.com/openwrt/openwrt/issues/20298, https://github.com/openwrt/openwrt/issues/20206

namiltd commented:
arch_align_stack() function before patching:
```
8000425c <arch_align_stack>:
8000425c:	27bdffe8 	addiu	sp,sp,-24
80004260:	afb00010 	sw	s0,16(sp)
80004264:	afbf0014 	sw	ra,20(sp)
80004268:	8f820000 	lw	v0,0(gp)
8000426c:	8c420200 	lw	v0,512(v0)
80004270:	7c420480 	ext	v0,v0,0x12,0x1
80004274:	14400005 	bnez	v0,8000428c <arch_align_stack+0x30>
80004278:	00808025 	move	s0,a0
8000427c:	3c028078 	lui	v0,0x8078
80004280:	8c42549c 	lw	v0,21660(v0)
80004284:	14400007 	bnez	v0,800042a4 <arch_align_stack+0x48>
80004288:	00000000 	nop
8000428c:	8fbf0014 	lw	ra,20(sp)
80004290:	2402fff8 	li	v0,-8
80004294:	02021024 	and	v0,s0,v0
80004298:	8fb00010 	lw	s0,16(sp)
8000429c:	03e00008 	jr	ra
800042a0:	27bd0018 	addiu	sp,sp,24
800042a4:	0c0d8251 	jal	80360944 <get_random_u16> <------ jump
800042a8:	00000000 	nop
800042ac:	8fbf0014 	lw	ra,20(sp)
800042b0:	7c427900 	ext	v0,v0,0x4,0x10
800042b4:	02028023 	subu	s0,s0,v0
800042b8:	2402fff8 	li	v0,-8
800042bc:	02021024 	and	v0,s0,v0
800042c0:	8fb00010 	lw	s0,16(sp)
800042c4:	03e00008 	jr	ra
800042c8:	27bd0018 	addiu	sp,sp,24
```
and after:
```
8000425c <arch_align_stack>:
8000425c:	27bdffe0 	addiu	sp,sp,-32
80004260:	afb00018 	sw	s0,24(sp)
80004264:	afbf001c 	sw	ra,28(sp)
80004268:	8f820000 	lw	v0,0(gp)
8000426c:	8c420200 	lw	v0,512(v0)
80004270:	7c420480 	ext	v0,v0,0x12,0x1
80004274:	14400005 	bnez	v0,8000428c <arch_align_stack+0x30>
80004278:	00808025 	move	s0,a0
8000427c:	3c028078 	lui	v0,0x8078
80004280:	8c42349c 	lw	v0,13468(v0)
80004284:	14400007 	bnez	v0,800042a4 <arch_align_stack+0x48>
80004288:	3c028036 	lui	v0,0x8036
8000428c:	8fbf001c 	lw	ra,28(sp)
80004290:	2402fff8 	li	v0,-8
80004294:	02021024 	and	v0,s0,v0
80004298:	8fb00018 	lw	s0,24(sp)
8000429c:	03e00008 	jr	ra
800042a0:	27bd0020 	addiu	sp,sp,32
800042a4:	24420974 	addiu	v0,v0,2420
800042a8:	afa20014 	sw	v0,20(sp)
800042ac:	8fa20014 	lw	v0,20(sp)
800042b0:	0040f809 	jalr	v0  <------ jump
800042b4:	00000000 	nop
800042b8:	8fbf001c 	lw	ra,28(sp)
800042bc:	7c427900 	ext	v0,v0,0x4,0x10
800042c0:	02028023 	subu	s0,s0,v0
800042c4:	2402fff8 	li	v0,-8
800042c8:	02021024 	and	v0,s0,v0
800042cc:	8fb00018 	lw	s0,24(sp)
800042d0:	03e00008 	jr	ra
800042d4:	27bd0020 	addiu	sp,sp,32
```

KanjiMonster commented:
> > And more technical, your patch was broken, i.e tabs were converted to spaces, and one line was missing the initial `+`.
> 
> I'm sorry but I don't see spaces anywhere instead of tabs (except for one line where they are intentional), nor the missing +

https://patchwork.kernel.org/project/linux-mips/patch/825527059.1016043.1759968706607@mail.yahoo.com/ this is how your patch arrived on the list.

Note that there are no tabs (all spaces), and the line `#if defined(CONFIG_MIPS) && !defined(CONFIG_CPU_MIPSR6)` is broken, as its first character is neither `+`, `-` or a space.

Likewise the line after the `+    }` does not have a space at all.

This is what I get when I try to apply the patch:
```
applying: ramips/mt7620: IRQ-protect stack ASLR randomization in arch_align_stack
error: corrupt patch at line 4
Patch failed at 0001 ramips/mt7620: IRQ-protect stack ASLR randomization in arch_align_stack
hint: Use 'git am --show-current-patch=diff' to see the failed patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
```

neheb commented:
@zx2c4 there might be a cache issue honestly. reminds me of mt7621 having broken cache invalidation: https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?h=next-20251024&id=55a2aa08b3af519a9693f99cdf7fa6d8b62d9f65

namiltd commented:
The problem with this PR occurred on 24.10.4 - kernel v6.6. The problem with https://github.com/openwrt/openwrt/issues/20206 was only found on snapshot - kernel v6.12.

namiltd committed:
mt7620: workaround jal imm26 and redundant PAGE_ALIGN

On MT7620-class platforms (CONFIG_NET_RALINK_MT7620) we observe sporadic
wrong-jump-targets, kernel oopses, hanging, corrupted backtraces or even
"half-written" instructions when the compiler emits a direct 'jal imm26'
call.
This is triggered in:
  - the small random helpers inside get_random_u32_below(), and
  - the blkcg_maybe_throttle_current() call in resume_user_mode_work().

This patch forces those two call sites to use an indirect call via
a volatile function pointer (load into register + jalr) when building
for MT7620, avoiding embedding a 26-bit immediate jump target.

Additionally, on MT7620 builds the exec path in fs/exec.c is modified:
  - skip arch_align_stack() + PAGE_ALIGN() in setup_arg_pages()
    because the micro-randomization (< PAGE_SIZE) implemented by many
    ports (including MT7620) is negated immediately by PAGE_ALIGN().
    Skipping the redundant PAGE_ALIGN() reduces exposure to the
    problematic code pattern.

These changes are targeted workarounds for MT7620; behavioral logic is unchanged.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/20553
Signed-off-by: Robert Marko <robimarko@gmail.com>

namiltd commented:
Currently, PR increases stability on kernel 6.6 and 6.12. I think that until the cause is found, this workaround is the best solution.

joseluissanromatato commented:
Hello,

I've faced this similar issue on the mt7620 class platform provoking reboots.
```
[  334.701551] CPU 0 Unable to handle kernel paging request at virtual address 703614d8, epcc
[  334.712521] Oops[#1]:
[  334.714848] CPU: 0 PID: 6878 Comm: usbmode Tainted: G           O       6.6.104 #0
[  334.722582] $ 0   : 00000000 00000001 00000002 fffff000
[  334.727946] $ 4   : 7fa5b000 7fa5b000 00000001 0065646f
[  334.733300] $ 8   : 00000010 8061b02c 00000010 80808080
[  334.738654] $12   : fefefeff 00000001 00000001 ffffff00
[  334.744009] $16   : 7fa5b000 00000001 834ddb40 8580d600
[  334.749363] $20   : 8587cb40 8580d05c 837f61e0 00000001
[  334.754718] $24   : 00000000 ffffffff                  
[  334.760073] $28   : 8359a000 8359bd20 8359bde8 800042ac
[  334.765429] Hi    : 554a9555
[  334.768365] Lo    : 7ffaa000
[  334.771301] epc   : 703614d8 0x703614d8
[  334.775223] ra    : 800042ac arch_align_stack+0x50/0x70
[  334.780577] Status: 1100a403 KERNEL EXL IE 
[  334.784862] Cause : 50800008 (ExcCode 02)
[  334.788953] BadVA : 703614d8
[  334.791889] PrId  : 00019650 (MIPS 24KEc)
[  334.795979] Modules linked in: rt2800soc(O) rt2800mmio(O) rt2800lib(O) qcserial pppoe pppt
[  334.796649]  x_tables ip_set_list_set ip_set_hash_netportnet ip_set_hash_netport ip_set_hc
[  334.942094] Process usbmode (pid: 6878, threadinfo=0ef2b16c, task=6a331fd0, tls=77e81df4)
[  334.950453] Stack : 00001601 835553c0 00003561 80a01e64 8580d000 801ac798 834b4a80 8587cb0
[  334.959026]         80793280 00000001 8359bde8 80179798 00000000 00000000 00000000 0000000
[  334.967594]         00000000 00000000 00000000 00000000 00000001 834b4a80 00000000 8355530
[  334.976162]         00000001 80031368 8580d000 834b4a80 00000000 22ef8f0c ffffffff 8587cb0
[  334.984731]         8580d054 80149a5c 81827900 85841f00 00000100 80203b94 00000034 b47f443
[  334.993301]         ...
[  334.995803] Call Trace:
[  334.995808] 
[  334.999814] [<801ac798>] setup_arg_pages+0x48/0x2cc
[  335.004811] [<80179798>] free_unref_page+0x4c/0x124
[  335.009814] [<80031368>] flush_itimer_signals+0x34/0x5c
[  335.015159] [<80149a5c>] arch_pick_mmap_layout+0x1a4/0x1c4
[  335.020773] [<80203b94>] load_elf_phdrs+0x78/0xcc
[  335.025605] [<80205008>] load_elf_binary+0x2e0/0x1478
[  335.030774] [<801a3864>] __kernel_read+0x164/0x2ac
[  335.035688] [<801b19ac>] getname_kernel+0x34/0x100
[  335.040606] [<801acf54>] bprm_execve+0x1ec/0x560
[  335.045326] [<801ad764>] copy_string_kernel+0x110/0x250
[  335.050673] [<801ae2e8>] do_execveat_common+0x1b4/0x240
[  335.056016] [<8000fd54>] do_page_fault+0x1e4/0x554
[  335.060918] [<801af114>] sys_execve+0x34/0x48
[  335.065380] [<8000dd00>] syscall_common+0x34/0x58
[  335.070208] 
[  335.071726] Code: (Bad address in epc)
[  335.071726] 
[  335.077056] 
[  335.078699] ---[ end trace 0000000000000000 ]---
[  335.083421] Kernel panic - not syncing: Fatal exception
```
I've tested this patch and it works well for me on kernel 6.6.110 in version 24.10.4. I no longer experience reboots, and the device is now stable.
I'm happy to help with any follow-up needed to get it merged upstream.

@namiltd why it is not good enough to use it as an interim solution?

namiltd commented:
@joseluissanromatato On which device did these crashes occur?

ptpt52 commented:
tested days and no crash.

miyatsu commented:
Will this be merged before next major release? I have a Xiaomi mini (with ~~mt76220~~ mt7620 soc) can not boot after clean install of 24.10.5. Tried 21.02.7 works fine.

miyatsu commented:
> > Will this be merged before next major release? I have a Xiaomi mini (with ~mt76220~ mt7620 soc) can not boot after clean install of 24.10.5. Tried 21.02.7 works fine.
> 
> @miyatsu did you by chance try to compile OpenWRT with this patch applied? Does the Xiaomi min work for you then?

Never tried to compile openwrt my self. I can do some basic test on my Xiaomi mini device if you can provide me patched (and compiled) binary or download link for CI generated bin(s). ):

namiltd commented:
@robimarko Is there a chance for 24.10, the version on which tests were conducted the longest?
